contracts/external/cw-token-swap/src/types.rs

```diff
@@ -0,0 +1,420 @@
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{to_binary, Addr, BankMsg, Binary, Coin, CosmosMsg, Deps, Uint128, WasmMsg};
+
+use crate::ContractError;
+
+// We create empty trait and implement it for the types we want to use
+// so we can create a generic function that only accepts a vec of those types
+trait CompletionMsgsToCosmosMsg {
+    fn into_checked_cosmos_msg(
+        self,
+        deps: Deps,
+        denom: &str,
+    ) -> Result<(Uint128, CosmosMsg), ContractError>;
+}
+
+/// This function returns a vector of CosmosMsgs from the given vector of CompletionMsgs
+/// It verifies the msgs is not empty (else return empty vec)
+/// It verifies the total amount of funds matches the funds sent in all messages
+fn completion_to_cosmos_msgs<T: CompletionMsgsToCosmosMsg>(
+    deps: Deps,
+    msgs: Vec<T>,
+    amount: Uint128,
+    string: &str, // extra data we need (denom or contract addr)
+) -> Result<Vec<CosmosMsg>, ContractError> {
+    if msgs.is_empty() {
+        return Ok(vec![]);
+    }
+
+    let mut total_amount = Uint128::zero();
+    let cosmos_msgs = msgs
+        .into_iter()
+        .map(|msg| {
+            let (amount, cosmos_msg) = msg.into_checked_cosmos_msg(deps, string)?;
+            total_amount += amount;
+            Ok(cosmos_msg)
+        })
+        .collect::<Result<Vec<CosmosMsg>, ContractError>>()?;
+
+    // Verify that total amount of funds matches funds sent in all messages
+    if total_amount != amount {
+        return Err(ContractError::WrongFundsCalculation {});
+    }
+    Ok(cosmos_msgs)
+}
+
+#[cw_serde]
+pub enum NativeSendMsg {
+    BankSend {
+        to_address: String,
+        amount: Vec<Coin>,
+    },
+    BankBurn {
+        amount: Vec<Coin>,
+    },
+    WasmExecute {
+        contract_addr: String,
+        msg: Binary,
+        funds: Vec<Coin>,
+    },
+    WasmInstantiate {
+        admin: Option<String>,
+        code_id: u64,
+        msg: Binary,
+        funds: Vec<Coin>,
+        label: String,
+    },
+}
+
+impl CompletionMsgsToCosmosMsg for NativeSendMsg {
+    fn into_checked_cosmos_msg(
+        self,
+        deps: Deps,
+        denom: &str,
+    ) -> Result<(Uint128, CosmosMsg), ContractError> {
+        let verify_coin = |coins: &Vec<Coin>| {
+            if coins.len() != 1 {
+                return Err(ContractError::InvalidSendMsg {});
+            }
+            if coins[0].amount.is_zero() {
+                return Err(ContractError::InvalidSendMsg {});
+            }
+            if denom != coins[0].denom {
+                return Err(ContractError::InvalidSendMsg {});
+            }
+            Ok(coins[0].amount)
+        };
```

Comment:
i think this is a reimplementation of cw-utils [`must_pay`](https://github.com/CosmWasm/cw-utils/blob/main/src/payment.rs#L30-L39). might be good to use that here.

Reply:
You are not wrong, but `must_pay` accepts `MessageInfo` and from the looks of it was built specifically for testing funds sent with msgs, which is not the case here (Exactly, kind of).
To use that we either need to:
1. build `MessageInfo` struct with unchecked addr, on every match.
2. or create 1 `mut MessageInfo` struct and modify the funds field on every match.

Both feels hackish and wrong to me, I think having this incapsulated helper function makes it clear what it does and why its there.

Reply:
great point :) thanks.